### DISCUSSION TITLE
How to handle same entity on different schemas

### DISCUSSION THREAD

**Discussion by Nahuel92** [micronaut-projects/micronaut-data#3010] [Q&A] [answered]
Hi there!

I'm using Micronaut Data JDBC and I have the following situation:

There is an entity that exists in more than one DB schema. Since the entity's definition is the same one among schemas, I would like to have a unique `JdbcRepository` that allows me to pass a parameter to determine which schema to use at runtime. I'm not sure this can be done at this time.

A workaround I could use is to create a base entity with all the fields and several entities extending the base one and annotated with `@MappedEntity(value = "my_common_entity", schema = "schema_name")` with their corresponding repositories.

Another workaround is to use `JdbcOperations` and do what I need all by hand, but I'm trying to see if there is a better approach for this.

I'll appreciate any suggestions on this. Thanks in advance!

**Comment by dstepanov** [comment]
Some databases support defining the schema in the context. Maybe we can support overriding the schema in `JdbcRepository`

**Reply by Nahuel92** [reply]
Hi dstepanov.

Yes, if I could pass the schema as a method parameter at runtime that would be awesome. I'm using Postgres, by the way.

**Reply by dstepanov** [reply]
For postgres you can simply change the current default schema  https://www.postgresql.org/docs/9.3/sql-set.html

**Reply by Nahuel92** [reply]
I forgot to add that I'll need to switch at least between 5 schemas, in case it helps.

**Reply by dstepanov** [reply]
Use an sql query to switch schema

**Reply by dstepanov** [reply]
Multi tenancy already does it this way

**Reply by Nahuel92** [reply]
Got it, I'll check the docs to find an example to do that and will give it a try. Thanks!!

**Reply by Nahuel92** [reply]
@dstepanov Do you happen to have an example I could use as a guide?

I understand how to include a Tenant in my entity, but what I still don't completely understand is how to reuse the same repository and make it to pick the tenant to choose the right schema.

If it helps, here's the (simplified) code I have and the SQL DDLs:

### Entity

```java
@Serdeable
@MappedEntity(value = "my_shared_entity")
public class MySharedEntity {
    @Id
    @MappedProperty("id")
    private UUID id;

    @MappedProperty("first_name")
    private String firstName;

    @TenantId
    private String tenant;
}
```

### Repository

```java
@JdbcRepository(dialect = Dialect.POSTGRES)
public interface MyRepository extends CrudRepository<MySharedEntity, UUID> {
}
```

### DDL

```sql
-- First schema & table
CREATE SCHEMA IF NOT EXISTS schema1;
CREATE TABLE IF NOT EXISTS schema1.my_shared_entity
(
    id UUID      NOT NULL PRIMARY KEY,
    first_name                       VARCHAR
);

-- Second schema & table
CREATE SCHEMA IF NOT EXISTS schema2;
CREATE TABLE IF NOT EXISTS schema2.my_shared_entity
(
    id UUID      NOT NULL PRIMARY KEY,
    first_name                       VARCHAR
);

-- Third schema & table
CREATE SCHEMA IF NOT EXISTS schema3;
CREATE TABLE IF NOT EXISTS schema3.my_shared_entity
(
    id UUID      NOT NULL PRIMARY KEY,
    first_name                       VARCHAR
);
```

Thanks!!!

**Reply by dstepanov** [reply]
Please check some of the multitenancy schema examples in the repository

**Reply by Nahuel92** [reply]
Thanks, that helped me a bit.

Unfortunately, I believe none of the built-in `TenantResolvers` can help me as I'm working on an app with multiple cron jobs (I don't expose a REST API, so we don't get called). Mine in particular calls a third-party API via a unique endpoint, and this endpoint responds with a DTO that target our different DB schemas (it includes an attribute that specifies which schema to use).

To make things worse, I can't use `JdbcSchemaHandler` (as I saw in the repo examples) to change the schema because it seems it changes the schema globally, and the other cron jobs that run in parallel use the same DB connection to read/write data (it's the DBA policy to only allow one connection per app and I can't change it).

I'll go over the examples again to see if I missed anything that can help me and I'm not seeing yet.

**Reply by dstepanov** [reply]
Well you need to change the schema, do your work and change it back

**Reply by Nahuel92** [reply]
Yes, I thought of that too, but I don't think it's possible to do so on my particular project. Let me give you more context:

The app I'm working on has (so far) 3 cron-job methods (annotated with `@Scheduled`) that run in different threads (and each uses a virtual thread executor to further parallelize processing). Each "pipeline", if you will, has a bunch of repositories for different entities (and from different schemas) injected.

On the other hand, I can get the schema from the `Connection`, but if I modify the schema and change it back via `JdbcSchemaHandler`, wouldn't that cause errors in other repositories executing code during the schema changing process? I know my situation is very complex, but I can't do much about it.

On a different subject, I tried setting `tenant` manually in the entity, but it doesn't work as I expected because Micronaut Data can't find the table at runtime. Using above's entity and repository:

```java
final var entity = new MySharedEntity();
entity.setTenant("schema1");
myRepository.save(entity);

final var entity2 = new MySharedEntity();
entity2.setTenant("schema2");
myRepository.save(entity2);

final var entity3 = new MySharedEntity();
entity3.setTenant("schema3");
myRepository.save(entity3);
```

Previous code causes Micronaut to error with `io.micronaut.data.exceptions.DataAccessException: Error executing PERSIST: ERROR: relation "my_shared_entity" does not exist`.

Not sure if that is an allowed use of tenancy. but I wish it is, as it could save me a lot of pain and code. I'd love to hear from you any other suggestion. Anyway, I really appreciate your help and patience.

**Reply by dstepanov** [reply]
The connection is only shared after your job is done with so if you reset the schema it should be fine

**Reply by Nahuel92** [reply]
Interesting, I didn't know that. In that case, you solved my problem then (I already have a PoC, I was just waiting to hear from you on the job/threading concern).

Is there any documentation/place that states what you mentioned? I would like to document my code with it for future developers. If not, I'll link your comment above. 

I'm planning to use a lock to protect method calls to `MyRepository` because I'll use virtual threads to process each DTO in parallel. My concern is about any race condition that may change the schema in the wrong moment and cause sneaky bugs.

**Reply by dstepanov** [reply]
Well that the behaviour of every connection pool. If you modify any kind of state there (read only etc) you need to put the previous state after you are done with it. Once again, for every connection work you can a connection assigned as after it's being put back into the pool.

What do you mean by lock protect methods? I don't think it's needed.

**Reply by Nahuel92** [reply]
I forgot to mention that my connection pool has only one connection (hard rule by our DBA). I'll review how a connection pool works, thanks.

Regarding the lock to protect methods calls to `MyRepository`, I meant this. Suppose my cron job calls an endpoint that returns this payload:

```json
[
  {
    "id": "1",
    "schema": "schema1",
    "firstName": "name1"
  },
  {
    "id": "2",
    "schema": "schema3",
    "firstName": "name2"
  },
  {
    "id": "3",
    "schema": "schema2",
    "firstName": "name3"
  },
  {
    "id": "4",
    "schema": "schema1",
    "firstName": "name4"
  }
]
```

For each one, I'm creating a virtual thread to process it in parallel.

My Java code looks like this:

```java
// constructor omitted for brevity
@Singleton
class MyCronJob {
    private final MyService myService;

    @Scheduled(fixedRate = "10s")
    void run() {
        myService.process();
    }
}
```

```java
// constructor omitted for brevity
@Singleton
class MyService { 
    private final MyDeclarativeApi myDeclarativeApi;
    private final MyMapper myMapper;
    private final MyRepository myRepository;
    private final Connection connection;
    private final JdbcSchemaHandler jdbcSchemaHandler;

    public void process() {
        List<MyDto> dtos = myDeclarativeApi.getAll();

        List<Runnable> entitiesToProcess = dtos.stream()
                .map(e -> createRunnable())
                .toList();

        try (var executor = Executors.newVirtualThreadPerTaskExecutor()) {
            entitiesToProcess.forEach(e -> executor.submit(e));
        }
    }
    
    private Runnable createRunnable() {
        return dto -> {
            String originalSchema;
            try {
                 originalSchema = connection.getSchema();
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
            jdbcSchemaHandler.useSchema(connection, Dialect.POSTGRES, dto.getSchema()); // consider schema valid at this point

            MySharedEntity mySharedEntity = myMapper.map(dto);
            MySharedEntity existingEntity = myRepository.findById(mySharedEntity.getId());
            if (existingEntity.isPresent()) {
                myRepository.update(mySharedEntity);
            } else {
                myRepository.save(mySharedEntity);
            }
            jdbcSchemaHandler.useSchema(connection, Dialect.POSTGRES, originalSchema);
        };
    }
}
```

I'm concerned that most of the `Runnable` code needs to be protected by a Lock to avoid any race condition (such as DTO 1 changing schema for DTO 3, for instance). I appreciate any input on this. Thanks a lot :)

**Comment by Nahuel92** [comment, marked as answer]
Hey, I wanted to give an update on this discussion.

In the end, I opted for using `JdbcOperations` and pass it a query created dynamically by hand (after validating parameters to avoid a SQL-injection vulnerability). This is because schemas can't be passed as a parameter to `PreparedStatement`. It's not ideal but it saves me from more complexity and code duplication that, otherwise, I would have ended up with.

It would be awesome if I could use Micronaut Data JDBC to map the same entity to different schemas, but as far as I know it seems not to be possible (I couldn't, at least).
